### PR TITLE
[MNT] Migrate from `pykalman` to `pykalman-bardo`

### PR DESCRIPTION
**Pull request by mbalatsko**
Consider migration from `pykalman` to `pykalman-bardo`, as [pykalman](https://github.com/pykalman/pykalman) project is no longer maintained. There were some issues that were fixed, see: https://github.com/pybardo/pykalman/blob/v0.9.6/CHANGELOG

I'm going to maintain [pykalman-bardo](https://github.com/pybardo/pykalman), react to issues, and fix bugs. For now the API is the same as in the initial package, but it might evolve in time.

I hope you will find it useful for your project!

#### Reference Issues/PRs
`pykalman` compatibility with python 3.11

#### Does your contribution introduce a new dependency? If yes, which one?
`pykalman`-> `pykalman-bardo`

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [ ] Optionally, I've added myself and possibly others to the [CODEOWNERS](https://github.com/sktime/sktime/blob/main/CODEOWNERS) file - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [ ] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [ ] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [ ] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [ ] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

### CODE REVIEW

**Review by yarnabrina — CHANGES_REQUESTED** [review]
Thanks for the PR, much appreciated.

However, this is insufficient to handle `pykalman` dependencies everywhere. You need to modify `python_dependencies` tag to updated package names, and also have to add the tag `python_dependencies_alias` which tracks import name of the packages.

At least this needs to be changed:

https://github.com/sktime/sktime/blob/d4a0dc3f9a622b673bc385a83a8271f5fe99b81c/sktime/transformations/series/kalman_filter.py#L536

File: pyproject.toml
Comment (mbalatsko) [reply pyproject.toml:103]: I would keep it as a defensive bound as 0.10 might introduce slight API changes

**Comment by mbalatsko** [comment]
> Thanks for the PR, much appreciated.
> 
> However, this is insufficient to handle `pykalman` dependencies everywhere. You need to modify `python_dependencies` tag to updated package names, and also have to add the tag `python_dependencies_alias` which tracks import name of the packages.
> 
> At least this needs to be changed:
> 
> https://github.com/sktime/sktime/blob/d4a0dc3f9a622b673bc385a83a8271f5fe99b81c/sktime/transformations/series/kalman_filter.py#L536

@yarnabrina Good point! I have greped the whole codebase and replaced `pykalman` -> `pykalman-bardo` where it is necessary. Do you think it looks sufficient now?

File: pyproject.toml
Comment (yarnabrina) [reply pyproject.toml:103]:
```suggestion
    "pykalman-bardo>=0.9.7,<0.10",
```

So, this should be fine I suppose?

(FYI @fkiraly as we usually disagree on lower bound upgrade points.)

File: sktime/transformations/series/kalman_filter.py
Comment (yarnabrina) [inline sktime/transformations/series/kalman_filter.py:536]:
```suggestion
        "python_dependencies": "pykalman-bardo",
        "python_dependencies_alias": {"pykalman-bardo": "pykalman"},
```

I think this is necessary.

File: sktime/transformations/series/tests/test_kalman_filter.py
Comment (yarnabrina) [inline sktime/transformations/series/tests/test_kalman_filter.py:203]:
```suggestion
    not _check_soft_dependencies("pykalman-bardo", package_import_alias={"pykalman-bardo": "pykalman"}, severity="none"),
```

I think this is required, plus black formatting.

File: sktime/transformations/series/tests/test_kalman_filter.py
Comment (yarnabrina) [inline sktime/transformations/series/tests/test_kalman_filter.py:348]:
```suggestion
    not _check_soft_dependencies("pykalman-bardo", "filterpy", package_import_alias={"pykalman-bardo": "pykalman"}, severity="none"),
```

File: sktime/transformations/series/tests/test_kalman_filter.py
Comment (yarnabrina) [inline sktime/transformations/series/tests/test_kalman_filter.py:583]: Why remove `pykalman` part? 😕

Comment (yarnabrina) [reply sktime/transformations/series/tests/test_kalman_filter.py:583]:
```suggestion
    not _check_soft_dependencies("pykalman-bardo", "filterpy", package_import_alias={"pykalman-bardo": "pykalman"}, severity="none"),
```

**Review by yarnabrina — CHANGES_REQUESTED** [review]
I think these changes are required as well. But I'll trigger CI just to see if anything else fails.

**Comment by mbalatsko** [comment]
@yarnabrina could we please try trigger CI again? I believe I have fixed all mentioned issues and codestyle

**Comment by mbalatsko** [comment]
Looks like one docstring was broken, fixed

**Comment by fkiraly** [comment]
welcome back from the dead, `pykalman`!
@mbalatsko, it would appear we just missed each other in Prague.

Regarding `pykalman`, are you aware of PEP 541 requests?
If a package is no longer maintained and the maintainers are no longer responsive, the python software foundation can hand the pypi name to you. Details at the bottom.

I assume this would apply in this case, as it seems like PR on the `pykalman` repo went unanswered for years.
If you were to become the owner of `pykalman` proper on `pypi`, it would make things easier.

Although there's no clear indication on the time frame, it is at least the "attemting contact" period long which is weeks.

Make PEP 541 requests here: https://github.com/pypi/support/issues
Details: https://peps.python.org/pep-0541/

**Comment by fkiraly** [comment]
FYI @NoaBenAmi as the original author of the `pykalman` interface

**Comment by mbalatsko** [comment]
> welcome back from the dead, `pykalman`! @mbalatsko, it would appear we just missed each other in Prague.
> 
> Regarding `pykalman`, are you aware of PEP 541 requests? If a package is no longer maintained and the maintainers are no longer responsive, the python software foundation can hand the pypi name to you. Details at the bottom.
> 
> I assume this would apply in this case, as it seems like PR on the `pykalman` repo went unanswered for years. If you were to become the owner of `pykalman` proper on `pypi`, it would make things easier.
> 
> Although there's no clear indication on the time frame, it is at least the "attemting contact" period long which is weeks.
> 
> Make PEP 541 requests here: https://github.com/pypi/support/issues Details: https://peps.python.org/pep-0541/

Oooh, that's cool, I was not aware of it. Let me submit the PR there! 

But as you said, it could take a while. Do you think that python 3.11 is something that you support now, or it can wait until I take over the PyPI project? In other words, do you prefer to wait until the takeover, or you'd prefer to finish and merge this PR and later I woud make `pykalman-bardo` an alias of `pykalman` on PyPI, and submit a new PR to this repo migrating back to `pykalman`?

**Comment by fkiraly** [comment]
> But as you said, it could take a while. Do you think that python 3.11 is something that you support now, or it can wait until I take over the PyPI project? 

Hm, that's an interesting question.

* the earliest the dependency switch can happen is with the 0.24.0 release, as we do not want to break our users' downstream setups. We just released 0.23.0, so 0.24.0 is likely mid to late Oct.
* so, I would suggest, let's finish this PR so it is ready to go with 0.24.0, as a release specific PR. If you were successful with the transfer by then, we close this PR. If not, we merge this PR with 0.24.0, and switch back with the first minor release after the transfer.

**Comment by mbalatsko** [comment]
@fkiraly soft requirements issue should be fixed, local tests passing sucessfully

**Comment by fkiraly** [comment]
FYI @yarnabrina, in order to avoid downstream breakage, I think we should only merge at 0.24.0 release.

**Comment by fkiraly** [comment]
Here's another thing that comes to my mind: @noabenami looked into `pykalman` a while ago, and she encountered limitations so had to supplement and mix/match functionality with `filterpy`. I do not remember the details anymore, but perhaps there was something that is relevant to plan the tech roadmap - @NoaBenAmi, are yous still following messages? Your notes in the `pykalman` / `filterpy` topic might be useful.

**Comment by fkiraly** [comment]
hm, the PEP request seems stuck - so let's switch to `pykalman-bardo` for now with 0.24.0, then switch back at the next minor after the proper handover (I assume this is max 6 weeks from now as the maintainer of `pykalman` has indicated they are happy to hand over, but might have disengageg from discussion, so we may have to wait until PSF transfers the pypi name)

**Review by fkiraly — APPROVED** [review]
scheduled for 0.24.0

**Pull request merged**
